clash-prelude/src/Clash/Class/Finite/Internal.hs

```diff
@@ -0,0 +1,751 @@
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-# OPTIONS_GHC -fplugin GHC.TypeLits.Normalise #-}
+{-# OPTIONS_GHC -fplugin GHC.TypeLits.KnownNat.Solver #-}
+{-# OPTIONS_GHC -fplugin GHC.TypeLits.Extra.Solver #-}
+{-# OPTIONS_GHC -fconstraint-solver-iterations=20 #-}
+
+module Clash.Class.Finite.Internal where
+
+import Prelude hiding ((++), (!!), concatMap, foldl, foldr, repeat, reverse)
+
+import Control.Applicative (Alternative(..))
+import Control.Arrow (second)
+import Data.Bits (Bits(..), FiniteBits(..))
+import Data.Coerce (coerce)
+import Data.Constraint (Dict(..))
+import Data.Functor.Compose (Compose(..))
+import Data.Functor.Const (Const(..))
+import Data.Functor.Identity (Identity(..))
+import Data.Functor.Product (Product)
+import Data.Functor.Sum (Sum)
+import Data.Int (Int8, Int16, Int32, Int64)
+import Data.Kind (Type)
+import Data.Maybe (fromMaybe, fromJust)
+import Data.Ord (Down(..))
+import Data.Proxy (Proxy(..))
+import Data.Singletons (Apply, TyFun)
+import Data.Void (Void)
+import Data.Word (Word8, Word16, Word32, Word64)
+import GHC.Generics
+  (Generic(..), Rep, V1, U1(..), M1(..), K1(..), (:+:)(..), (:*:)(..))
+import GHC.TypeNats
+  (Nat, KnownNat, type (^), type (<=), type (*), type (+), type (-))
+
+import Clash.Class.Num (SaturatingNum(..), SaturationMode(..))
+import Clash.Class.Finite.Internal.Dictionaries
+  ( powGTZero, powMonotone1, powLawsRewrite
+#if !MIN_VERSION_base(4,16,0)
+  , powCLogDual
+#endif
+  )
+import Clash.Class.Finite.Internal.TH (deriveFiniteTuples)
+import Clash.Class.Resize (Resize(..))
+import Clash.Class.BitPack (BitPack(..), bitCoerce)
+import Clash.Promoted.Nat
+  ( SNat(..), UNat(..), toUNat, fromUNat, natToNum, snatToNum
+  )
+import Clash.Sized.Index (Index)
+import Clash.Sized.Internal.BitVector (BitVector(..), Bit(..), high, low)
+import Clash.Sized.Internal.Unsigned (Unsigned(..))
+import Clash.Sized.Signed (Signed)
+import Clash.Sized.RTree (RTree(..), tdfold, tfold, trepeat)
+import Clash.Sized.Vector
+  ( Vec(..), (++), (!!), concatMap, bv2v, dfold, foldl, foldr
+  , ifoldr, indicesI, iterateI, unfoldrI, repeat, reverse, replace
+  )
+import Clash.XException (ShowX, NFDataX)
+
+import qualified Data.List as List (iterate)
+
+{- $setup
+>>> :m -Prelude
+>>> :set -XDeriveAnyClass
+>>> import Clash.Prelude
+-}
+
+-- * Finite Class
+
+-- | The class of types holding only a finite number of elements. The
+-- class offers type level access to the number of elements @n@ and
+-- defines a total order on the elements via indexing them from @0@ to
+-- @n-1@. The class gives access to the vector of all elements of the
+-- type, allows to iterate over the elements in order or to map back
+-- and forth between their associated indices.
+--
+-- In general, the class can be considered as a more hardware-friendly
+-- alternative to 'Bounded' and 'Enum', utilizing 'Index' instead of
+-- 'Int' and vectors instead of lists.
```

Comment:
Is there a reason that we don't provide `FiniteEnum` (or `CEnum`) directly? This would give developers an immediate intuition for what the classes are.

Related: I've wondered many times whether it would be sensible for `Clash.Prelude` to export a Clash-friendly version of all these `Int`/`Integer` based APIs. I.e., replace `base`'s `Enum` (and friends) in favor of Clash ones.

Reply:
I would not change the semantics of the classes that are already available in `base`, because this only leads to trouble and confusion. The right strategy should be to provide hardware-friendy alternatives, like `Signed n` is to be preferred over `Int` in hardware, which then can coexist with the software-friendly types.

Most of the design choices for this extension have been inspired by my [finite](https://hackage.haskell.org/package/finite-1.4.1.2/docs/Finite.html) library for sure, where I considered it more useful to have a single class covering all of the properties of a "finite" type, instead of breaking things up into multiple classes like `Enum` or `Bounded`, which definitely makes sense for types with infinitely many inhabitants.

I'd also like to avoid to make the interface here too similar to `Enum` and friends, because there still are some fundamental differences, which could be mixed up easily otherwise.